### PR TITLE
Doc: fix linalg.lstsq documentation on residues

### PR DESCRIPTION
**Pull request by f380cedric**
The condition on residues is on the rank of A, not the number of dimension of A.

### CODE REVIEW

**Comment by melissawm** [comment]
Hi @f380cedric - so sorry for the long wait on a review here! 

This does look confusing, because initially the docstring would mention rank and it was changed to ndim [in this commit](https://github.com/scipy/scipy/pull/9888/files#diff-8400a871af5e3ff38643aff36ff192650313e769cdfe80f57b02140def686ee1). This was probably a mistake, as indeed looking at, for example, [dgelsd](https://netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga94bd4a63a6dacf523e25ff617719f752.html#ga94bd4a63a6dacf523e25ff617719f752), the documentation for `B` says

```
B is DOUBLE PRECISION array, dimension (LDB,NRHS)
          On entry, the M-by-NRHS right hand side matrix B.
          On exit, B is overwritten by the N-by-NRHS solution
          matrix X.  If m >= n **and RANK = n**, the residual
          sum-of-squares for the solution in the i-th column is given
          by the sum of squares of elements n+1:m in that column.
```

As far as your second question, the relevant code seems to be [this block](https://github.com/scipy/scipy/blob/7221501dc2ba4763505fa8e153bb4c256381f0c5/scipy/linalg/_basic.py#L1211) which does list the condition `m>n` so I think the docstring is correct.

**Comment by f380cedric** [comment]
> Hi @f380cedric - so sorry for the long wait on a review here!

Thx for reviewing this :)

> As far as your second question, the relevant code seems to be [this block](https://github.com/scipy/scipy/blob/7221501dc2ba4763505fa8e153bb4c256381f0c5/scipy/linalg/_basic.py#L1211) which does list the condition `m>n` so I think the docstring is correct.

Yeah, the dostring is definitely representing the behavior of the code.
However, I was questioning the code itself because `gelsd` says `m >= n`, while scipy says `m > n`.
After thinking about it, `m = n && rank(A) = n` should mean that residuals are 0s,
so returning them is useless.

File: scipy/linalg/_basic.py [hunk 1]
```diff
@@ -1081,7 +1081,7 @@ def lstsq(a, b, cond=None, overwrite_a=False, overwrite_b=False,
         Least-squares solution.
     residues : (K,) ndarray or float
         Square of the 2-norm for each column in ``b - a x``, if ``M > N`` and
-        ``ndim(A) == n`` (returns a scalar if ``b`` is 1-D). Otherwise a
+        ``rank(A) == n`` (returns a scalar if ``b`` is 1-D). Otherwise a
```
Comment (h-vetinari) [inline scipy/linalg/_basic.py:1084]: So I think this should actually be "if M >= N and rank(A) == N".

**Review by h-vetinari — COMMENTED** [review]
Can we get this buttoned up for 1.11 still? :)

> > As far as your second question, the relevant code seems to be [this block](https://github.com/scipy/scipy/blob/7221501dc2ba4763505fa8e153bb4c256381f0c5/scipy/linalg/_basic.py#L1211) which does list the condition `m>n` so I think the docstring is correct.
> 
> Yeah, the dostring is definitely representing the behavior of the code.

I don't actually agree here. The branch pointed out by Melissa is only taken for `m>n`, but the code still runs through for m==n. Also, LAPACK docs for dgelsd explicitly [say](https://github.com/Reference-LAPACK/lapack/blob/v3.6.0/SRC/dgelsd.f#L42-L43) (going back for many versions):
> A is an M-by-N matrix which may be rank-deficient.

It also [says](https://github.com/Reference-LAPACK/lapack/blob/v3.6.0/SRC/dgelsd.f#L110-L112):
> If m >= n and RANK = n, the residual sum-of-squares for the solution in the i-th column is given by the sum of squares of elements n+1:m in that column.

PS. Everything here is horribly inconsistent w.r.t. capitalization - i.e. the signature calls the matrix `a` and then describes `Ax=b` at first, before falling back to `b - ax` etc.

**Comment by f380cedric** [comment]
> > > As far as your second question, the relevant code seems to be [this block](https://github.com/scipy/scipy/blob/7221501dc2ba4763505fa8e153bb4c256381f0c5/scipy/linalg/_basic.py#L1211) which does list the condition `m>n` so I think the docstring is correct.
> > 
> > 
> > Yeah, the dostring is definitely representing the behavior of the code.
> 
> I don't actually agree here. The branch pointed out by Melissa is only taken for `m>n`, but the code still runs through for m==n.

That part of the docstring is only regarding the shape of the returned residuals. From my understanding, the code run for `m==n` (as it should), but the shape of residual is `!= (0,)` only if `rank(A) == n && m > n`,
which is in agreement with the docstring.

> Also, LAPACK docs for dgelsd explicitly [say](https://github.com/Reference-LAPACK/lapack/blob/v3.6.0/SRC/dgelsd.f#L42-L43) (going back for many versions):
> 
> > A is an M-by-N matrix which may be rank-deficient.
> 
> It also [says](https://github.com/Reference-LAPACK/lapack/blob/v3.6.0/SRC/dgelsd.f#L110-L112):
> 
> > If m >= n and RANK = n, the residual sum-of-squares for the solution in the i-th column is given by the sum of squares of elements n+1:m in that column.

If `m == n`, then `n+1:m` is an empty list. So you have non-empty residuals iff `rank(A) == n && m > n`.

> PS. Everything here is horribly inconsistent w.r.t. capitalization - i.e. the signature calls the matrix `a` and then describes `Ax=b` at first, before falling back to `b - ax` etc.

I can push something for that,
should I uppercase or lowercase them?

File: scipy/linalg/_basic.py (hunk 1 above)
Comment (tylerjereddy) [reply scipy/linalg/_basic.py:1084]: @ilayn do you agree on this suggestion?

**Comment by tylerjereddy** [comment]
This obviously isn't critical for the release, but I did ping one of our linear algebra experts to see if we can sort out this one-line change for your first contribution sooner than later.

**Comment by tylerjereddy** [comment]
Oof, we're past 1 year of review on a 1-line doc change. Unfortunately, the discussion above is still far to nebulous for me to confidently merge something here.

Let me check if @ilayn has an opinion here, but I'll bump the milestone in the interim. The next release (`1.13.0`) after this one (`1.12.0`) may be a little faster than the usual cycle, so there may be an opportunity to push the change out soon-ish anyway, if a consensus can be reached.

**Comment by tylerjereddy** [comment]
@ev-br you've been touching linalg stuff lately, any thoughts on how to resolve this discussion around a 1-line/4-character change in a docstring?

Seems a bit silly to keep bumping the milestone on this one. I think I'm just going to remove the milestone given how small/non-critical the change is. If this gets merged a milestone can be re-added...

**Comment by ev-br** [comment]
Checking the code, it seems to be indeed the (matrix) rank of A which decides, not the dimension of A:

https://github.com/scipy/scipy/blob/v1.12.0/scipy/linalg/_basic.py#L1295

and the `rank` variable comes from the LAPACK call (gelss or gelssd), where it is also documented as "The effective rank of A, i.e., the number of singular values which are greater than RCOND*S(1)."

https://www.netlib.org/lapack/explore-html/da/d55/group__gelss_gac6159de3953ae0386c2799294745ac90.html#gac6159de3953ae0386c2799294745ac90

While the logic of the code is quite baroque, it's way too late to change it in backwards incompatible ways. 
This PR is an improvement, and further tweaks to the docstring can be done in a sequel. Thanks @f380cedric , all!

**Pull request merged**